### ISSUE TITLE
New rootdisksize not set with Scale instance in 4.17.2

### ISSUE THREAD

**Issue by ccycv** [apache/cloudstack#7325]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
4.17.1 and 4.17.2
~~~

##### CONFIGURATION

~~~
Advanced networking with local storage. vmware 6.7.
~~~

##### OS / ENVIRONMENT

~~~
N/A
~~~
##### Service Offering used on the instance
~~~
{
  "count": 1,
  "serviceoffering": [
    {
      "cacheMode": "none",
      "cpuspeed": 1600,
      "created": "2022-11-16T08:21:16+0000",
      "defaultuse": false,
      "diskofferingstrictness": false,
      "displaytext": "S-Packages",
      "dynamicscalingenabled": true,
      "hasannotations": false,
      "id": "08f5f8fa-88d7-44c7-8ed5-7f06a6031888",
      "iscustomized": true,
      "issystem": false,
      "isvolatile": false,
      "limitcpuuse": false,
      "name": "S-Packages",
      "networkrate": 500,
      "offerha": false,
      "provisioningtype": "thin",
      "rootdisksize": 0,
      "serviceofferingdetails": {
        "maxcpunumber": "8",
        "maxmemory": "32768",
        "mincpunumber": "1",
        "minmemory": "512"
      },
      "storagetype": "local"
    }
  ]
}
~~~
##### Virtualmachine

~~~
{
  "count": 1,
  "virtualmachine": [
    {
      "account": "emeaclient_16_49",
      "affinitygroup": [],
      "cpunumber": 1,
      "cpuspeed": 1600,
      "created": "2023-03-07T10:21:04+0000",
      "details": {
        "Message.ReservedCapacityFreed.Flag": "true",
        "cpuNumber": "1",
        "cpuOvercommitRatio": "2.0",
        "cpuSpeed": "1600",
        "dataDiskController": "osdefault",
        "keyboard": "us",
        "memory": "1024",
        "memoryOvercommitRatio": "2.0",
        "nicAdapter": "Vmxnet3",
        "rootDiskController": "scsi",
        "rootdisksize": "10"
      },
      "displayname": "Test-Scale-March-3",
      "displayvm": true,
      "domain": "cloud16",
      "domainid": "793d17f0-30d3-46c1-adbf-657c4659a624",
      "guestosid": "6428ffbd-1f58-11ec-b6e8-244bfeb8116f",
      "haenable": false,
      "hasannotations": false,
      "hypervisor": "VMware",
      "id": "a2b9aca8-4e67-4cb4-8526-e2507b8d3898",
      "instancename": "i-74-1443-VM",
      "isdynamicallyscalable": false,
      "lastupdated": "2023-03-07T10:30:43+0000",
      "memory": 1024,
      "name": "VM-a2b9aca8-4e67-4cb4-8526-e2507b8d3898",
      "nic": [
        {
          "broadcasturi": "vlan://untagged",
          "deviceid": "0",
          "extradhcpoption": [],
          "gateway": "46.xx.xxx.49",
          "id": "4e23f2be-ecc5-4793-8cbd-c3bc15920cd1",
          "ipaddress": "46.xx.xxx.54",
          "isdefault": true,
          "isolationuri": "vlan://untagged",
          "macaddress": "1e:00:cf:00:00:87",
          "netmask": "255.255.255.240",
          "networkid": "3f5fbbe4-fe38-42ae-af2d-01709d7a2e9c",
          "networkname": "NL-01",
          "secondaryip": [],
          "traffictype": "Guest",
          "type": "Shared"
        }
      ],
      "osdisplayname": "Other Linux (64-bit)",
      "ostypeid": "6428ffbd-1f58-11ec-b6e8-244bfeb8116f",
      "passwordenabled": true,
      "pooltype": "VMFS",
      "receivedbytes": 0,
      "rootdeviceid": 0,
      "rootdevicetype": "ROOT",
      "securitygroup": [],
      "sentbytes": 0,
      "serviceofferingid": "08f5f8fa-88d7-44c7-8ed5-7f06a6031888",
      "serviceofferingname": "S-Packages",
      "state": "Stopped",
      "tags": [
        {
          "account": "emeaclient_16_49",
          "domain": "cloud16",
          "domainid": "793d17f0-30d3-46c1-adbf-657c4659a624",
          "key": "service_id",
          "resourceid": "a2b9aca8-4e67-4cb4-8526-e2507b8d3898",
          "resourcetype": "UserVm",
          "value": "450"
        },
        {
          "account": "emeaclient_16_49",
          "domain": "cloud16",
          "domainid": "793d17f0-30d3-46c1-adbf-657c4659a624",
          "key": "service_type",
          "resourceid": "a2b9aca8-4e67-4cb4-8526-e2507b8d3898",
          "resourcetype": "UserVm",
          "value": "S-2"
        }
      ],
      "templatedisplaytext": "RockyLinux-8",
      "templateid": "be405fbf-9ca9-4b90-bf97-9806ebf75b7b",
      "templatename": "RockyLinux-8",
      "userid": "df0acf1e-52a2-4c4f-b4e6-535546cf6b12",
      "username": "emeaclient_16_49",
      "zoneid": "55f0123b-285c-4a70-8b34-565d8454393d",
      "zonename": "NL-01"
    }
  ]
}
~~~

##### SUMMARY

New value for “rootdisksize” parameter does not have any effect on the instance

##### STEPS TO REPRODUCE

~~~
scale virtualmachine id=a2b9aca8-4e67-4cb4-8526-e2507b8d3898 serviceofferingid=08f5f8fa-88d7-44c7-8ed5-7f06a6031888 details[0].cpuNumber=1 details[0].memory=1024 details[0].rootdisksize=25
~~~

##### EXPECTED RESULTS

~~~
New value for rootdisk to be set, and not ignored, 


~~~

##### ACTUAL RESULTS

~~~
Only the CPU and Memory is scaled, rootdisksize is ignored without any error in the log.

GET  apiKey=fXfTV81qwEJ_9WrhJYtYOXE3oPSLlEjfNZx2i9YLUMZD1Zw61_PySJv3ethOdvQUBIFrvMI6UYwQJZtvNNp-Eg&command=scaleVirtualMachine&details[0].cpuNumber=1&details[0].memory=1024&details[0].rootdisksize=25&id=a2b9aca8-4e67-4cb4-8526-e2507b8d3898&response=json&serviceofferingid=08f5f8fa-88d7-44c7-8ed5-7f06a6031888&signature=A1d99N328ZK2SbfDo%2FP74xMpPcU%3D
~~~

**Comment by weizhouapache** [comment]
@ccycv 
You scaled the vm to same custom service offering which has custom disk offering , right ?

**Comment by ccycv** [comment]
> @ccycv You scaled the vm to same custom service offering which has custom disk offering , right ?

I scale the VM with the same service offering. The VM was created only with root disk with the above service offering, no extra disk offering.

**Comment by kiranchavala** [comment]
@ccycv 

Did you check global setting "allow.diskoffering.change.during.scale.vm"

**Comment by ccycv** [comment]
> @ccycv
> 
> Did you check global setting "allow.diskoffering.change.during.scale.vm"

Yes, I only forgot to mention here.

"I have also changed the value to "true" for allow.diskoffering.change.during.scale.vm. The only difference is that I do not see any more the warning in the log "Changing the disk offering of the root volume during the compute offering change operation is disabled. Please check the setting [allow.diskoffering.change.during.scale.vm]""

**Comment by kiranchavala** [comment]
@ccycv 


I see that the "rootdisksize": 0, for the service offering id  "08f5f8fa-88d7-44c7-8ed5-7f06a6031888"

Could you please create custom service offering with root disk say 20 gb  and then try scaling of the vm 




Ideally, resizeVolume api will be triggered as part for the  scale virtual machine job

 https://cloudstack.apache.org/api/apidocs-4.17/apis/resizeVolume.html
 
My lab  Logs

~~~

2023-03-09 06:36:33,520 DEBUG [c.c.a.t.Request] (Work-Job-Executor-7:ctx-0d678fa0 job-193/job-194 ctx-1616eb76) (logid:38b3bcf8) Seq 2-236720455413664548: Sending  { Cmd , MgmtId: 32989174039340, via: 2(ref-trl-4565-k-M7-kiran-chavala-kvm2), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.storage.ResizeVolumeCommand":{"path":"ab2b387c-4d2d-4e12-9ec0-b64a9406eb74","pool":{"id":"1","uuid":"93ae92ce-5e8c-3460-b256-d0162835d9fe","host":"10.0.32.4","path":"/acs/primary/ref-trl-4565-k-M7-kiran-chavala/ref-trl-4565-k-M7-kiran-chavala-kvm-pri1","port":"2049","type":"NetworkFilesystem"},"currentSize":"8589934592","newSize":"(20.00 GB) 21474836480","shrinkOk":"false","vmInstance":"i-2-20-VM","managed":"false","wait":"0","bypassHostMaintenance":"false"}}] }
~~~


Could you please send the management server logs ( /var/log/cloudstack/management/management-server.log )

**Comment by weizhouapache** [comment]
> allow.diskoffering.change.during.scale.vm

@ccycv 
what's the value in zone setting ?

**Comment by ccycv** [comment]
> > allow.diskoffering.change.during.scale.vm
> 
> @ccycv what's the value in zone setting ?

"true" as I mentioned above.

**Comment by ccycv** [comment]
> > allow.diskoffering.change.during.scale.vm
> 
> @ccycv what's the value in zone setting ?

everywhere is true. I did this check before.  There is no error returned via API or in log with debug enabled.

**Comment by weizhouapache** [comment]
> @weizhouapache @kiranchavala
> 
> I have found additional info
> 
> "Volume Vol[1515|vm=1443|ROOT] is already having disk offering ebfec3b0-5685-4a0f-b5eb-175b754b3a64"
> 
@ccycv 
I think this explains: The disk offering of the service offering is not customized. You can use resizeVolume .

In additional to @kiranchavala 's testing, I have tested the following steps
1. create a custom disk offering
2. create a contraint service offering, which uses the disk offering created in step 1.
3. create a vm with Small offering (or Custom offering with compute-only disk offering)
4. stop the vm and scale the vm to new offering created in step 2. root disk is resized.
5. scale the vm again with same offering but new root size. root disk is not resized. got an log: `Scaling CPU and/or Memory of VM with custom service offering. New disk offering stills the same. Skipping the Disk offering change on Volume 8501cd62-28b7-4042-bd24-dfcb7653c49a.`, which seems to be expected (see #6267)

**Comment by ccycv** [comment]
> > @weizhouapache @kiranchavala
> > I have found additional info
> > "Volume Vol[1515|vm=1443|ROOT] is already having disk offering ebfec3b0-5685-4a0f-b5eb-175b754b3a64"
> 
> @ccycv I think this explains: The disk offering of the service offering is not customized. You can use resizeVolume .
> 
> In additional to @kiranchavala 's testing, I have tested the following steps
> 
> 1. create a custom disk offering
> 2. create a contraint service offering, which uses the disk offering created in step 1.
> 3. create a vm with Small offering (or Custom offering with compute-only disk offering)
> 4. stop the vm and scale the vm to new offering created in step 2. root disk is resized.
> 5. scale the vm again with same offering but new root size. root disk is not resized. got an log: `Scaling CPU and/or Memory of VM with custom service offering. New disk offering stills the same. Skipping the Disk offering change on Volume 8501cd62-28b7-4042-bd24-dfcb7653c49a.`, which seems to be expected (see [Fix #6263 Cannot scale VM with custom offering #6267](https://github.com/apache/cloudstack/pull/6267))

FYI. I tested with service offering which use a custom disk offering and not, the results are the same.  The message I see in the log "Volume Vol[1515|vm=1443|ROOT] is already having disk offering ebfec3b0-5685-4a0f-b5eb-175b754b3a64" Is for the service offering which does not use any disk offering, and this disk offering "ebfec3b0-5685-4a0f-b5eb-175b754b3a64"does not exist.

**Comment by kiranchavala** [comment]
hi @ccycv 

I also got the same message as @weizhouapache 

1. Create a custom compute offering , uncheck the compute-only disk offering and select a custom disk offering

![compute offering](https://user-images.githubusercontent.com/1401014/224026598-a0773ed4-d770-4a58-b236-a301759719c4.png)

2. deploy a vm using the custom compute offering, by default it will take the minimum root disk size as 8 gb 

![deployvm](https://user-images.githubusercontent.com/1401014/224027166-c2cb6c70-cf98-4115-89a7-f2c4b9bd095c.png)

3. Stop the vm 

4. Scale the vm to the same cutom offering but mention the new disk size of 25 gb

![scale](https://user-images.githubusercontent.com/1401014/224029086-ac70a994-c9c1-4eea-a400-bc3ce9cba0e0.png)


5. The volume will not be scaled  and the following logs 

~~~
2023-03-09 12:35:22,281 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-43:ctx-c544c5f6 job-289) (logid:179580d3) Executing AsyncJobVO: {id:289, userId: 2, accountId: 2, instanceType: VirtualMachine, instanceId: 30, cmd: org.apache.cloudstack.api.command.admin.vm.ScaleVMCmdByAdmin, cmdInfo: {"details[0].cpuNumber":"1","httpmethod":"GET","ctxAccountId":"2","uuid":"ab277894-4527-47ea-8c34-eecbf726ea13","automigrate":"true","details[0].memory":"512","cmdEventType":"VM.UPGRADE","serviceofferingid":"58c06adf-db66-442d-ba26-8ea06230949f","response":"json","ctxUserId":"2","ctxStartEventId":"1002","id":"ab277894-4527-47ea-8c34-eecbf726ea13","ctxDetails":"{\"interface com.cloud.vm.VirtualMachine\":\"ab277894-4527-47ea-8c34-eecbf726ea13\",\"interface com.cloud.offering.ServiceOffering\":\"58c06adf-db66-442d-ba26-8ea06230949f\"}","details[0].rootdisksize":"25"}, cmdVersion: 0, status: IN_PROGRESS, processStatus: 0, resultCode: 0, result: null, initMsid: 32989174039340, completeMsid: null, lastUpdated: null, lastPolled: null, created: null, removed: null}


2023-03-09 12:35:22,300 DEBUG [c.c.u.AccountManagerImpl] (API-Job-Executor-43:ctx-c544c5f6 job-289 ctx-f3a9d065) (logid:179580d3) Access granted to Account [{"accountName":"admin","id":2,"uuid":"1011d1ed-b73a-11ed-87f7-1e00e400032c"}] to Service offering {"id":24,"name":"c1","uuid":null}. by AffinityGroupAccessChecker


2023-03-09 12:35:22,304 DEBUG [c.c.s.VolumeApiServiceImpl] (API-Job-Executor-43:ctx-c544c5f6 job-289 ctx-f3a9d065) (logid:179580d3) Scaling CPU and/or Memory of VM with custom service offering. New disk offering stills the same. Skipping the Disk offering change on Volume 7bfa9c06-c87f-450d-89b5-de9b96bef20d.

~~~

6. you have to select the root disk volume and resize to particular size

![resize](https://user-images.githubusercontent.com/1401014/224029564-3ca2491c-98de-449e-83f2-e576f2364fa7.png)

**Comment by ccycv** [comment]
> hi @ccycv
> 
> I also got the same message as @weizhouapache
> 
> 1. Create a custom compute offering , uncheck the compute-only disk offering and select a custom disk offering
> 
> ![compute offering](https://user-images.githubusercontent.com/1401014/224026598-a0773ed4-d770-4a58-b236-a301759719c4.png)
> 
> 2. deploy a vm using the custom compute offering, by default it will take the minimum root disk size as 8 gb
> 
> ![deployvm](https://user-images.githubusercontent.com/1401014/224027166-c2cb6c70-cf98-4115-89a7-f2c4b9bd095c.png)
> 
> 3. Stop the vm
> 4. Scale the vm to the same cutom offering but mention the new disk size of 25 gb
> 
> ![scale](https://user-images.githubusercontent.com/1401014/224029086-ac70a994-c9c1-4eea-a400-bc3ce9cba0e0.png)
> 
> 5. The volume will not be scaled  and the following logs
> 
> ```
> 2023-03-09 12:35:22,281 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-43:ctx-c544c5f6 job-289) (logid:179580d3) Executing AsyncJobVO: {id:289, userId: 2, accountId: 2, instanceType: VirtualMachine, instanceId: 30, cmd: org.apache.cloudstack.api.command.admin.vm.ScaleVMCmdByAdmin, cmdInfo: {"details[0].cpuNumber":"1","httpmethod":"GET","ctxAccountId":"2","uuid":"ab277894-4527-47ea-8c34-eecbf726ea13","automigrate":"true","details[0].memory":"512","cmdEventType":"VM.UPGRADE","serviceofferingid":"58c06adf-db66-442d-ba26-8ea06230949f","response":"json","ctxUserId":"2","ctxStartEventId":"1002","id":"ab277894-4527-47ea-8c34-eecbf726ea13","ctxDetails":"{\"interface com.cloud.vm.VirtualMachine\":\"ab277894-4527-47ea-8c34-eecbf726ea13\",\"interface com.cloud.offering.ServiceOffering\":\"58c06adf-db66-442d-ba26-8ea06230949f\"}","details[0].rootdisksize":"25"}, cmdVersion: 0, status: IN_PROGRESS, processStatus: 0, resultCode: 0, result: null, initMsid: 32989174039340, completeMsid: null, lastUpdated: null, lastPolled: null, created: null, removed: null}
> 
> 
> 2023-03-09 12:35:22,300 DEBUG [c.c.u.AccountManagerImpl] (API-Job-Executor-43:ctx-c544c5f6 job-289 ctx-f3a9d065) (logid:179580d3) Access granted to Account [{"accountName":"admin","id":2,"uuid":"1011d1ed-b73a-11ed-87f7-1e00e400032c"}] to Service offering {"id":24,"name":"c1","uuid":null}. by AffinityGroupAccessChecker
> 
> 
> 2023-03-09 12:35:22,304 DEBUG [c.c.s.VolumeApiServiceImpl] (API-Job-Executor-43:ctx-c544c5f6 job-289 ctx-f3a9d065) (logid:179580d3) Scaling CPU and/or Memory of VM with custom service offering. New disk offering stills the same. Skipping the Disk offering change on Volume 7bfa9c06-c87f-450d-89b5-de9b96bef20d.
> ```
> 
> 6. you have to select the root disk volume and resize to particular size
> 
> ![resize](https://user-images.githubusercontent.com/1401014/224029564-3ca2491c-98de-449e-83f2-e576f2364fa7.png)

Right, this is correct. This is how it should work? or it is a bug?   Yes, I know that I can resize the volume separately, but this requires an extra API call. Because i'm doing all these via API, I must list the volume to get the ID of it and then resize, instead just doing via scale.

**Comment by kiranchavala** [comment]
@ccycv  this is expected behaviour according to the code fix in #6267

@weizhouapache  can we consider this issue for an enhancement request

**Comment by harikrishna-patnala** [comment]
I see an issue here, in any case if linked diskoffering is custom then it should consider the new size while changing the compute offering.

**Comment by weizhouapache** [comment]
@harikrishna-patnala 
did you mean the check in #6267 is wrong ? (I personally agree..)

**Comment by harikrishna-patnala** [comment]
Yes @weizhouapache , we can skip only in the case having same non-custom diskofferings

**Comment by weizhouapache** [comment]
> Yes @weizhouapache , we can skip only in the case having same non-custom diskofferings

great we reache a consensus. @harikrishna-patnala @kiranchavala 

I have marked this as bug and added it to 4.18.1.0 milestone.

**Comment by shwstppr** [comment]
Fixed with #7359 
@ccycv you may test the latest nightly
